### PR TITLE
Add option for class_attribute default

### PR DESCRIPTION
**Pull request by dhh**
It's extremely common, as evidenced by the commit in this PR of Rails conversions, to set a default value for a class_attribute. So let's provide that directly.

### CODE REVIEW

File: actionview/lib/action_view/layouts.rb [hunk 1]
```diff
@@ -204,9 +204,9 @@ module Layouts
     include ActionView::Rendering
 
     included do
-      class_attribute :_layout, :_layout_conditions, instance_accessor: false
-      self._layout = nil
-      self._layout_conditions = {}
+      class_attribute :_layout
```
Comment (matthewd) [inline actionview/lib/action_view/layouts.rb]: Lost `instance_accessor: false`

File: activesupport/lib/active_support/core_ext/class/attribute.rb [hunk 2]
```diff
@@ -123,6 +124,10 @@ def class_attribute(*attrs)
         remove_possible_method "#{name}="
         attr_writer name
       end
+
+      if default_value
```
Comment (matthewd) [inline activesupport/lib/active_support/core_ext/class/attribute.rb]: `unless default_value.nil?`

Comment (matthewd) [inline activesupport/lib/active_support/core_ext/class/attribute.rb]: .. or just change the `nil` on line 79/80

File: actionpack/lib/action_controller/metal/helpers.rb [hunk 3]
```diff
@@ -53,9 +53,8 @@ class << self; attr_accessor :helpers_path; end
     include AbstractController::Helpers
 
     included do
-      class_attribute :helpers_path, :include_all_helpers
-      self.helpers_path ||= []
```
Comment (dhh) [inline actionpack/lib/action_controller/metal/helpers.rb]: Probably shouldn't replace this with a default:, right? Given that it checks the parent class first.

File: activesupport/lib/active_support/callbacks.rb [hunk 4]
```diff
@@ -62,8 +62,7 @@ module Callbacks
 
     included do
       extend ActiveSupport::DescendantsTracker
-      class_attribute :__callbacks, instance_writer: false
-      self.__callbacks ||= {}
```
Comment (dhh) [inline activesupport/lib/active_support/callbacks.rb]: Same issue here: "Probably shouldn't replace this with a default:, right? Given that it checks the parent class first."

File: actionpack/lib/action_controller/metal/helpers.rb (hunk 3 above)
Comment (matthewd) [inline actionpack/lib/action_controller/metal/helpers.rb]: I hesitated at that, but ultimately convinced myself the `||=` is a lie: `class_attribute` will always define a method on the current class, with the value set to nil... so AFAICS, this is no different from if it were a straight assignment.

Comment (dhh) [inline actionpack/lib/action_controller/metal/helpers.rb]: Ah, right. Yup, good 👍

File: activesupport/lib/active_support/core_ext/class/attribute.rb [hunk 5]
```diff
@@ -70,9 +70,10 @@ class Class
   # To opt out of both instance methods, pass <tt>instance_accessor: false</tt>.
   def class_attribute(*attrs)
     options = attrs.extract_options!
-    instance_reader = options.fetch(:instance_accessor, true) && options.fetch(:instance_reader, true)
-    instance_writer = options.fetch(:instance_accessor, true) && options.fetch(:instance_writer, true)
+    instance_reader    = options.fetch(:instance_accessor, true) && options.fetch(:instance_reader, true)
+    instance_writer    = options.fetch(:instance_accessor, true) && options.fetch(:instance_writer, true)
     instance_predicate = options.fetch(:instance_predicate, true)
+    default_value      = options.fetch(:default, nil)
```
Comment (matthewd) [inline activesupport/lib/active_support/core_ext/class/attribute.rb]: This (really, the fact it's in the `options` hash at all) does mean the methods below will capture the default value in their closure -- so even if the value is changed on the root class, the original default won't be GCed.

I don't think we particularly need to care about a couple of extra empty hashes and arrays floating around, but I figured I'd at least mention it.

Comment (dhh) [inline activesupport/lib/active_support/core_ext/class/attribute.rb]: Would be nice to convert all these to kwargs. Should probably do a big sweep for Rails 6.

File: activesupport/lib/active_support/core_ext/class/attribute.rb [hunk 6]
```diff
@@ -70,9 +70,10 @@ class Class
   # To opt out of both instance methods, pass <tt>instance_accessor: false</tt>.
   def class_attribute(*attrs)
     options = attrs.extract_options!
-    instance_reader = options.fetch(:instance_accessor, true) && options.fetch(:instance_reader, true)
-    instance_writer = options.fetch(:instance_accessor, true) && options.fetch(:instance_writer, true)
+    instance_reader    = options.fetch(:instance_accessor, true) && options.fetch(:instance_reader, true)
```
Comment (simi) [inline activesupport/lib/active_support/core_ext/class/attribute.rb]: *Sorry for a little off-topic here: I have spotted this formatting changes in few recent PRs. Is this new standard? Wasn't this considered "cosmetic" change before?*

**Comment by dhh** [comment]
Not a general new standard. Just a dash of BDFL privilege 😂

**Comment by gsamokovarov** [comment]
David, have you tried defaulting with a block?

```ruby
class_attribute :periodic_timers, instance_reader: false, default: []
```

vs

```ruby
class_attribute(:periodic_timers, instance_reader: false) { [] }
```

The reason I'm bringing it up is because `mattr_accessor` works like 👆 and does not have the `default` option. Should we pair the functionality so it's symmetrical across all the attribute extensions?

**Comment by dhh** [comment]
That strikes me as less revealing, but the upside is that it won't get evaluated until you access it, which is a nice perk. Default does have symmetry with belongs_to/default, though.

**Comment by gsamokovarov** [comment]
Add a `default` to `mattr_accessor` as well then? Dunno if it will make the interface too big though.

**Comment by dhh** [comment]
Other aspect is that class_attribute already has other named options. mattr_accessor doesn't and follows the rest of the Ruby methods by same name.

File: activesupport/lib/active_support/core_ext/class/attribute.rb (hunk 5 above)
Comment (rafaelfranca) [inline activesupport/lib/active_support/core_ext/class/attribute.rb]: We can do it right now. We only support Ruby versions that work with kwargs

**Comment by dhh** [comment]
Let's make it so!

**Comment by gsamokovarov** [comment]
[`mattr_accessor` family actually accept options as well](https://github.com/rails/rails/blob/6847877a30fd8d578f72db0cf40674f71a9b6286/activesupport/lib/active_support/core_ext/module/attribute_accessors.rb#L67), they let you choose whether or not to have the instance methods. The interfaces between `class_attributes` and `cattr_accessor` aren't actually that far off.

**Comment by kaspth** [comment]
> Let's make it so!

@dhh GitHub didn't connect your email reply: this was in reference to kwargs across the board right?

**Comment by dhh** [comment]
Kasper, yes.

**Comment by dhh** [comment]
True. I'd rather make mattr_accessor use default:, then, than going with
{}. For lazy resolution, we could always have default take a lambda.

**Comment by kaspth** [comment]
Looks like `mattr_*` just yields to the block immediately, so there's no lazy loading at all: https://github.com/rails/rails/blob/3fbe657e9b65814b3837fd13628e7a812dc0a0ea/activesupport/lib/active_support/core_ext/module/attribute_accessors.rb#L74

Then there doesn't seem to be a need for `default:` to take lambdas — besides all our `class_attribute` didn't need deferred evaluation, and I haven't seen the need for it in my app's usage of `mattr_*` either.

I'd be 👍 on deprecating the block to `mattr_*` and `cattr_*` and replacing it with a `default:` arg.

**Comment by dhh** [comment]
Ah, yeah, then I definitely don't think it's worth it. A kwarg seems
clearly better to me.

**Comment by gsamokovarov** [comment]
Cool, I can take stab at that.

**Pull request merged**